Add optional max_tokens argument for function splitByRegexp

taiyang-li opened the pull request:
### Changelog category (leave one):
- Improvement



### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Refactor FunctionTokens to enable max tokens returned for related functions(default disabled)


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

taiyang-li committed:
enable limits for functions using FunctionTokens

taiyang-li committed:
fix comment

vdimir commented on tests/queries/0_stateless/01866_split_by_regexp.reference:
```diff
@@ -5,3 +5,8 @@
+['one','two','three','']
+['one','two','three','']
+['one']
+['one','two']
```
Why `three` is omitted, not `['one','twoBthreeC']` ?

vdimir commented:
So, basically this PR `Add support of max_tokens argument for function splitByRegexp` ?

taiyang-li commented:
@vdimir  Yes. The original goal of this pr is to support max_tokens argument for splitByRegexp, just like in Spark.  Then I found it is better to implement max_tokens limitation in `FunctionTokens`,  which covers all related split-family functions. 

But I met one issue,  `splitByChar` has already supported `max_tokens` in Python semantics.  But I want to implement max_tokens for splitByChar in Spark semantics(for this project: https://github.com/oap-project/gluten).  What do you advice to handle the difference between two semantics?

taiyang-li commented:
Yes, you are right. In python, the argument is max split. 
```
>>> a = '11.12.13' 
>>> a.split('.', 3)  
['11', '12', '13']   # split at most 3 times, but there is only 3 tokens, return 3 tokens 
>>> a.split('.', 2)
['11', '12', '13']   # split at most 2 times, return 3 tokens
>>> a.split('.', 1)
['11', '12.13']     #  split at most 1 times, return 2 tokens.
>>> a.split('.', 0)
['11.12.13']        # not split, return the whole string. 
```

In Spark, the argument defines max tokens to return.

taiyang-li commented:
Now I think it is better to only add max_tokens argument for CH function `splitByRegexp(str[, max_tokens])`.  The refaction of `FunctionTokens` will be reverted.

vdimir commented:
>  Now I think it is better to add max_tokens argument only for CH function splitByRegexp(str[, max_tokens]).

Sounds good.

taiyang-li commented:
> I misunderstood `max_split` in `splitByChar` before. Now I think it is better to add max_tokens argument only for CH function `splitByRegexp(str[, max_tokens])`. The refaction of `FunctionTokens` will be reverted. Do you think it is ok ?

It is done.  @vdimir

vdimir commented:
Hi @taiyang-li 
I've checked PR once again. I'm sorry, but I was confused too. Now I realize that `splitByChar` accepts `maxsplit`. So it's better to have `maxsplit` for `spitByRegex` to be consistent. So probably your first version was almost correct?

And another point is that I suppose we shouldn't split the tail even if it matches the regex, because this behavior can be achieved with split + arrayResize.

```sql
SELECT splitByRegexp('[0-9]+', 'a123b456c7d', 2)


┌─splitByRegexp('[0-9]+', 'a123b456c7d', 2)─┐
│ ['a','b']                                 │
└───────────────────────────────────────────┘
```

In python it returns `b456c7d` :

```python
re.split('[0-9]+', 'a123b456c7d', 1)
Out[16]: ['a', 'b456c7d']
```

taiyang-li commented:
The behavior of splitByRegexp depends on whether we follow the rules of Python or Spark.  
If we follow the rules of Spark, current implementation is right. Otherwise the change you requested is right.

vdimir commented:
> The behavior of splitByRegexp depends on whether we follow the rules of Python or Spark.
If we follow the rules of Spark, current implementation is right. Otherwise the change you requested is right.

Yes, if splitByChar already follows python's rules, so let's make it the same.

taiyang-li commented:
Sorry, but what ourself really need is Spark style. Do you think improve splitByRegexp in Python style and adding a new function named `split` mapping to Spark built-in function `split` is a good idea?

rschu1ze commented on src/Functions/FunctionsStringArray.h:
```diff
@@ -493,11 +510,16 @@ class SplitByRegexpImpl
+        /// Returned tokens exceeds max_tokens limitation.
+        if (max_tokens && *max_tokens && curr_tokens >= *max_tokens)
+            return false;
+
```
Sort of unrelated: the check `if (!re)` (l. 523) could be done at the end of the `init()` method (basically throw an exception ... More importantly, the only reason for `!re` is an out-of-memory/bad_alloc, an invalid regexp will throw an exception already (see the `OptimizedRegularExpressionImpl` ctor in Common/OptimizedRegularExpression.cpp). But a bad_alloc is also an exception, so we might remove `if (!re)` altogether.

rschu1ze commented on docs/en/sql-reference/functions/splitting-merging-functions.md:
```diff
@@ -91,20 +91,21 @@ SELECT splitByString('', 'abcde');
 
 -   `regexp` — Regular expression. Constant. [String](../data-types/string.md) or [FixedString](../data-types/fixedstring.md).
 -   `s` — The string to split. [String](../../sql-reference/data-types/string.md).
+-   `max_tokens` — An optional `Int64` defaulting to 0. When `max_tokens` > 0, the returned tokens will be no more than `max_tokens`, otherwise `splitByRegexp` will return as many tokens as possible.
```
Ah, I should have read the previous discussion in this PR ^^

taiyang-li replied:
I also thought it is more reasonable to return empty arrays if input 0. But in spark function `split`, 0 means no limit. Since user can directly use empty string instead of  splitxxx(str, delim, 0),  can we just keep the same semantics with spark?

rschu1ze replied:
I don't mind if we go for Spark or Python semantics. What would be nice is if all `splitBy*()` functions had the same semantics for their 3rd parameter. And since the 3rd parameter of `splitByChar()` isn't documented, we are free to change it, imho.

taiyang-li commented on docs/en/sql-reference/functions/splitting-merging-functions.md:
````diff
@@ -91,20 +91,21 @@ SELECT splitByString('', 'abcde');
 
 ``` sql
-splitByRegexp(regexp, s)
+splitByRegexp(regexp, s[, max_tokens])
````
Thank you very much for your advance.  This is exactly the original change I have made, but I reverted it because it breaks  the semantics consistency of function `splitByChar`.   

Now I will add optional argument for all split-family functions.

taiyang-li commented on tests/queries/0_stateless/01866_split_by_regexp.sql:
```diff
@@ -3,3 +3,12 @@ select splitByRegexp('', 'abcde');
+select splitByRegexp('[ABC]', 'oneAtwoBthreeC', 2);
+select splitByRegexp('[ABC]', 'oneAtwoBthreeC', 3);
+select splitByRegexp('[ABC]', 'oneAtwoBthreeC', 4);
+select splitByRegexp('[ABC]', 'oneAtwoBthreeC', 5);
```
Good idea

taiyang-li committed:
Merge branch 'master' into enable_max_splits

taiyang-li committed:
reset to original solution

taiyang-li committed:
rename  alphaTokens to SplitByAlphaImpl

taiyang-li committed:
improve doc and uts

taiyang-li commented:
Now max_substrings is applied for alphaTokens/splitByNoApha/splitByWhitespace/splitByChar/splitByString/splitByRegexp

rschu1ze commented on docs/en/sql-reference/functions/splitting-merging-functions.md:
```diff
@@ -217,10 +226,28 @@ SELECT splitByNonAlpha('  1!  a,  b.  ');
 Returns the string.
 
-## alphaTokens(s)
+## alphaTokens(s[, max_substrings]), splitByAlpha(s[, max_substrings])
```
Thanks for fixing this up.

rschu1ze commented on src/Functions/FunctionsStringArray.h:
```diff
@@ -98,6 +98,12 @@ class AlphaTokensImpl
         return 0;
     }
 
+    /// Returns the position of the possible max_substrings argument. std::nullopt means max_substrings argument is disabled in current function.
```
No need to change anything, just as a hint: Having the positions (l. 96, l. 102) as static constants similar to name (l. 70) would have been a little bit more elegant.

taiyang-li replied:
Thanks. I agree with you. Maybe there is performance benefit after changing. Since currently we get string positions from static functions in FunctionTokens, let's keep max_substrings positions the same.

rschu1ze reviewed (approved):
Nicely done!

taiyang-li committed:
Merge branch 'master' into enable_max_splits

rschu1ze commented:
@taiyang-li There are still some issues with the functional tests:
- https://s3.amazonaws.com/clickhouse-test-reports/42673/e65b0f1c2e1d73b07748566efd59c94509dd8abd/fast_test.html
- https://s3.amazonaws.com/clickhouse-test-reports/42673/e65b0f1c2e1d73b07748566efd59c94509dd8abd/stateless_tests__aarch64_.html
- https://s3.amazonaws.com/clickhouse-test-reports/42673/e65b0f1c2e1d73b07748566efd59c94509dd8abd/stateless_tests__asan__[1/2].html
- https://s3.amazonaws.com/clickhouse-test-reports/42673/e65b0f1c2e1d73b07748566efd59c94509dd8abd/stateless_tests__release_.html
- https://s3.amazonaws.com/clickhouse-test-reports/42673/e65b0f1c2e1d73b07748566efd59c94509dd8abd/stateless_tests__release__s3_storage_.html
- https://s3.amazonaws.com/clickhouse-test-reports/42673/e65b0f1c2e1d73b07748566efd59c94509dd8abd/stateless_tests__release__wide_parts_enabled_.html
- https://s3.amazonaws.com/clickhouse-test-reports/42673/e65b0f1c2e1d73b07748566efd59c94509dd8abd/stateless_tests__ubsan_.html

Would you like to check them? Thanks!

taiyang-li committed:
Merge branch 'master' into enable_max_splits

taiyang-li committed:
Merge branch 'master' into enable_max_splits

rschu1ze commented:
@taiyang-li CI check "Fast Test" failed due to 00534_functions_bad_arguments2. This test was also red in previous runs of this PR, I actually suppose it is related to the changes. Would you like to check that? Thanks a lot!

taiyang-li commented:
@rschu1ze Sorry I can't reproduce it in my environment.  This is my local test of all functions that may be affected by the PR. 


```
 :) select splitByNonAlpha('')

SELECT splitByNonAlpha('')

Query id: 121e6970-92d9-48cd-87d6-c08cb2937583

┌─splitByNonAlpha('')─┐
│ []                  │
└─────────────────────┘

1 rows in set. Elapsed: 0.002 sec. 

 :) select splitByWhitespace('')

SELECT splitByWhitespace('')

Query id: 679f5268-a1c5-47d5-bf9c-3656e0b5e4da

┌─splitByWhitespace('')─┐
│ []                    │
└───────────────────────┘

1 rows in set. Elapsed: 0.002 sec. 

 :) select splitByChar('')

SELECT splitByChar('')

Query id: 94a2aa88-e2f1-44af-b3be-4a1852993d93


0 rows in set. Elapsed: 0.031 sec. 

Received exception from server (version 22.11.1):
Code: 42. DB::Exception: Received from localhost:9000. DB::Exception: Function 'splitByChar' needs at least 2 arguments, at most 3 arguments; passed 1.: While processing splitByChar(''). (NUMBER_OF_ARGUMENTS_DOESNT_MATCH)

 :) select splitByString('')

SELECT splitByString('')

Query id: 2772dadc-896e-4059-8fd6-5e984adaec3c


0 rows in set. Elapsed: 0.030 sec. 

Received exception from server (version 22.11.1):
Code: 42. DB::Exception: Received from localhost:9000. DB::Exception: Function 'splitByChar' needs at least 2 arguments, at most 3 arguments; passed 1.: While processing splitByString(''). (NUMBER_OF_ARGUMENTS_DOESNT_MATCH)

 :) select splitByRegexp('')

SELECT splitByRegexp('')

Query id: 6ba239d9-064b-4baa-b355-1af077b8adbf


0 rows in set. Elapsed: 0.030 sec. 

Received exception from server (version 22.11.1):
Code: 42. DB::Exception: Received from localhost:9000. DB::Exception: Function 'splitByChar' needs at least 2 arguments, at most 3 arguments; passed 1.: While processing splitByRegexp(''). (NUMBER_OF_ARGUMENTS_DOESNT_MATCH)

 :) select extractAll('')

SELECT extractAll('')

Query id: bb8f99ef-f48d-4680-8285-873edf54d61f


0 rows in set. Elapsed: 0.028 sec. 

Received exception from server (version 22.11.1):
Code: 42. DB::Exception: Received from localhost:9000. DB::Exception: Number of arguments for function extractAll doesn't match: passed 1, should be 2: While processing extractAll(''). (NUMBER_OF_ARGUMENTS_DOESNT_MATCH)

 :) select extractURLParameterNames('')

SELECT extractURLParameterNames('')

Query id: 8e709847-dbe6-4b29-b014-a7158aff1d1c

┌─extractURLParameterNames('')─┐
│ []                           │
└──────────────────────────────┘

1 rows in set. Elapsed: 0.002 sec. 

 :) select extractURLParameters('')

SELECT extractURLParameters('')

Query id: aec00bc4-a97f-4284-942b-b2016020a74b

┌─extractURLParameters('')─┐
│ []                       │
└──────────────────────────┘

1 rows in set. Elapsed: 0.002 sec. 

 :) select URLPathHierarchy('')

SELECT URLPathHierarchy('')

Query id: bc9e72c9-2798-4aae-a8e1-353fa16497c1

┌─URLPathHierarchy('')─┐
│ []                   │
└──────────────────────┘

1 rows in set. Elapsed: 0.001 sec. 

 :) select URLHierarchy('')

SELECT URLHierarchy('')

Query id: 8cbd0b3c-f52e-417b-adb7-a67cf6e90ddc

┌─URLHierarchy('')─┐
│ []               │
└──────────────────┘

1 rows in set. Elapsed: 0.001 sec. 

 :) Bye.

```

taiyang-li committed:
Merge branch 'master' into enable_max_splits

taiyang-li committed:
fix bug

taiyang-li committed:
Merge branch 'enable_max_splits' of https://github.com/bigo-sg/ClickHouse into enable_max_splits

taiyang-li commented:
> 

Sorry, it is fixed now.

taiyang-li committed:
Merge branch 'master' into enable_max_splits

aadant commented:
This change is rather brutal, it is introduced a regression compared to the previous behavior.

```
https://fiddle.clickhouse.com/314b915c-7c4d-436c-ad1b-a2276b66a561 - old version 
https://fiddle.clickhouse.com/283c0ca6-2055-4bc2-844c-926860c4380a - new version
```

rschu1ze commented:
@aadant Could you please share the brutal regression as SQL example? Fiddle died for unclear reason and it is unclear when it will come back.

rschu1ze commented:
It's alive again:

```sql
-- v22.8
select 'a=b=c=d' as _raw, splitByChar('=', _raw, 1); -- a=b=c=d	['a','b=c=d']
select 'a=b=c=d' as _raw, splitByChar('=', _raw, 2); -- a=b=c=d	['a','b','c=d']
select 'a=b=c=d' as _raw, splitByChar('=', _raw, 3); -- a=b=c=d	['a','b','c','d']

-- v23.3
select 'a=b=c=d' as _raw, splitByChar('=', _raw, 1); -- a=b=c=d	['a']
select 'a=b=c=d' as _raw, splitByChar('=', _raw, 2); -- a=b=c=d	['a','b']
select 'a=b=c=d' as _raw, splitByChar('=', _raw, 3); -- a=b=c=d	['a','b','c']
```

rschu1ze commented:
@aadant Kindly see the discussion following [this comment](https://github.com/ClickHouse/ClickHouse/pull/42673#issuecomment-1291832169). The previous "Pythonic" behavior was at no point documented, hence this can hardly be called a "regression".

aadant commented:
@rschu1ze : people's code relying on the old behavior is breaking. I call that a regression. The fact that it was not documented probably explains why it was done. But the documentation was missing and that itself was a documentation bug ! 

Maybe we could add an optional to fall back to the python behavior. There are probably more python users than spark users :-)

aadant commented:
I think we can just implement the split function and use the maxsplit argument. The code is what it was before ;-)

rschu1ze commented:
I am open towards a new setting that toggles Spark-ish vs. Pythonic behavior of the `max_substrings` arguments for the existing `splitBy*()` functions. But reverting the default behavior again (after changing it in this PR) or adding a new function `split` with Python semantics will maximally confuse users, so I'd be against that. I also think that only few users are affected by this PR at all (basically the ones who knew about the existence of the 3rd parameter by looking at the source code).

For now, allow me to update the documentation.

jhenhapl commented:
@rschu1ze I think another rather large regression here is that the tail of the string now gets split and trimmed. This doesn't follow _either_ Python or Spark semantics.

This was mentioned in https://github.com/ClickHouse/ClickHouse/pull/42673#issuecomment-1291832169 and https://github.com/ClickHouse/ClickHouse/pull/42673#issuecomment-1296928601 but still made it into the PR.

rschu1ze commented:
@aadant @jhenhapl @UnamedRus --> PR #54518
